File: sdks/akta-docs/java/src/main/java/dev/openakta/aktadocs/ChangelogAppend.java
package dev.openakta.aktadocs;

import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;

public final class ChangelogAppend {

    public static final String ANCHOR = "<!-- akta-changelog-append -->";

    private static final ObjectMapper JSON = new ObjectMapper();

    private ChangelogAppend() {}

    public record Result(String target, int bytesWritten, boolean created) {}

    public static Result append(Path target, String jsonPayload, String template, boolean dryRun)
            throws IOException {
        ChangelogEntryPayload entry = JSON.readValue(jsonPayload, ChangelogEntryPayload.class);
        String line =
                "- **"
                        + entry.changeType
                        + "** ("
                        + entry.timestamp
                        + ") "
                        + entry.summary;
        if ("detailed".equals(template) && entry.details != null && !entry.details.isBlank()) {
            line += "\n  " + entry.details.replace("\n", "\n  ");
        }
        String block = "\n" + line + "\n";

        String out;
        boolean created;
        if (!Files.exists(target)) {
            created = true;
            out =
                    "---\n"
                            + "doc_id: "
                            + entry.docId
                            + "\n"
                            + "doc_type: changelog\n"
                            + "date: "
                            + entry.timestamp.substring(0, Math.min(10, entry.timestamp.length()))
                            + "\n"
                            + "---\n\n"
                            + "# Changelog\n\n"
                            + ANCHOR
                            + block;
        } else {
            created = false;
            String existing = Files.readString(target);
            if (existing.contains(ANCHOR)) {
                out = existing.replace(ANCHOR, ANCHOR + block);
            } else {
                String sep = existing.endsWith("\n") ? "\n" : "\n\n";
                out = existing + sep + block.strip() + "\n";
            }
        }

        byte[] bytes = out.getBytes(java.nio.charset.StandardCharsets.UTF_8);
        if (dryRun) {
            return new Result(target.toString(), bytes.length, created);
        }
        Files.createDirectories(target.getParent());
        Path tmp = Files.createTempFile(target.getParent(), ".akta-changelog-", ".md");
        Files.write(tmp, bytes);
        Files.move(tmp, target, StandardCopyOption.REPLACE_EXISTING);
        return new Result(target.toString(), bytes.length, created);
    }
}

File: sdks/akta-docs/csharp/ChangelogAppend.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace OpenAkta.AktaDocs;

public sealed class ChangelogException : IOException
{
    public ChangelogException(string message, Exception? inner = null) : base(message, inner) { }
}

public sealed class ChangelogEntryPayload
{
    [JsonPropertyName("schema_version")]
    public string SchemaVersion { get; set; } = "";

    [JsonPropertyName("doc_id")]
    public string DocId { get; set; } = "";

    public string Timestamp { get; set; } = "";

    [JsonPropertyName("change_type")]
    public string ChangeType { get; set; } = "";

    public string Summary { get; set; } = "";

    public string? Details { get; set; }
}

public static class ChangelogAppend
{
    public const string Anchor = "<!-- akta-changelog-append -->";

    public sealed record Result(string Target, int BytesWritten, bool Created);

    public static Result Append(string targetPath, string jsonPayload, string template, bool dryRun)
    {
        ChangelogEntryPayload entry;
        try
        {
            entry = JsonSerializer.Deserialize<ChangelogEntryPayload>(jsonPayload)
                    ?? throw new ChangelogException("Invalid changelog payload: empty JSON.");
        }
        catch (JsonException e)
        {
            throw new ChangelogException("Invalid changelog payload: not valid JSON.", e);
        }

        if (string.IsNullOrWhiteSpace(entry.ChangeType) || string.IsNullOrWhiteSpace(entry.Summary))
            throw new ChangelogException("Invalid changelog payload: missing required fields.");

        var line = $"- **{entry.ChangeType}** ({entry.Timestamp}) {entry.Summary}";
        if (string.Equals(template, "detailed", StringComparison.Ordinal) &&
            !string.IsNullOrWhiteSpace(entry.Details))
            line += "\n  " + entry.Details!.Replace("\n", "\n  ", StringComparison.Ordinal);

        var block = "\n" + line + "\n";

        string outText;
        bool created;
        if (!File.Exists(targetPath))
        {
            created = true;
            var date = entry.Timestamp.Length >= 10 ? entry.Timestamp[..10] : entry.Timestamp;
            outText =
                "---\n" +
                "doc_id: " + entry.DocId + "\n" +
                "doc_type: changelog\n" +
                "date: " + date + "\n" +
                "---\n\n" +
                "# Changelog\n\n" +
                Anchor +
                block;
        }
        else
        {
            created = false;
            var existing = File.ReadAllText(targetPath);
            if (existing.Contains(Anchor, StringComparison.Ordinal))
                outText = existing.Replace(Anchor, Anchor + block, StringComparison.Ordinal);
            else
            {
                var sep = existing.EndsWith('\n') ? "\n" : "\n\n";
                outText = existing + sep + block.TrimStart() + "\n";
            }
        }

        var bytes = System.Text.Encoding.UTF8.GetBytes(outText);
        if (dryRun)
            return new Result(targetPath, bytes.Length, created);

        Directory.CreateDirectory(Path.GetDirectoryName(Path.GetFullPath(targetPath))!);
        var tmp = Path.Combine(Path.GetTempPath(), $"akta-changelog-{Guid.NewGuid():N}.md");
        try
        {
            File.WriteAllText(tmp, outText);
            File.Move(tmp, targetPath, overwrite: true);
        }
        finally
        {
            try
            {
                if (File.Exists(tmp)) File.Delete(tmp);
            }
            catch
            {
                /* ignore */
            }
        }

        return new Result(targetPath, bytes.Length, created);
    }
}
